File: app/src/main/java/com/vladimir_x/easyenglishlearn/exercises/ExerciseViewModel.java
package com.vladimir_x.easyenglishlearn.exercises;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.ViewModel;
import androidx.databinding.ObservableField;

import com.vladimir_x.easyenglishlearn.SingleLiveEvent;
import com.vladimir_x.easyenglishlearn.model.Answer;
import com.vladimir_x.easyenglishlearn.model.Word;

import java.util.List;

public abstract class ExerciseViewModel extends ViewModel {

    private SingleLiveEvent<Void> mExerciseCloseLiveData;
    private SingleLiveEvent<Integer> mMessageLiveData;
    private int mIteration;
    private int mErrorsCount;
    boolean mTranslationDirection;
    List<Word> mWordList;
    StringBuilder mAnswerBuilder;
    Word mCurrentWord;
    public final ObservableField<String> question = new ObservableField<>();


    ExerciseViewModel() {
        mAnswerBuilder = new StringBuilder();
        mExerciseCloseLiveData = new SingleLiveEvent<>();
        mMessageLiveData = new SingleLiveEvent<>();
    }

    /**
     * This method must be called from fragment  when it is first started
     */
    void startExercise(List<Word> wordList, boolean translationDirection) {
        mWordList = wordList;
        mTranslationDirection = translationDirection;
        mErrorsCount = 0;
        mIteration = 0;
        prepareQuestionAndAnswers();
    }

    void prepareQuestionAndAnswers() {
        mCurrentWord = mWordList.get(mIteration);
        question.set(mTranslationDirection
                ? mCurrentWord.getLexeme()
                : mCurrentWord.getTranslation());
        cleanPreviousData();
    }

    void checkAnswer() {
        if (isExerciseOver()) {
            finishExercise();
        } else {
            prepareQuestionAndAnswers();
        }
    }

    void cleanPreviousData() {
        mAnswerBuilder.delete(0, mAnswerBuilder.length());
    }

    private boolean isExerciseOver() {
        Answer answer = new Answer(mCurrentWord, mAnswerBuilder, mTranslationDirection);
        if (answer.isCorrect()) {
            return ++mIteration >= mWordList.size();
        } else {
            showMessage(-1);
            mErrorsCount++;
        }
        return false;
    }

    private void finishExercise() {
        showMessage(mErrorsCount);
        mExerciseCloseLiveData.call();
    }

    private void showMessage(int errorsCount) {
        mMessageLiveData.setValue(errorsCount);
    }

    LiveData<Void> getExerciseCloseLiveData() {
        return mExerciseCloseLiveData;
    }

    LiveData<Integer> getMessageLiveData() {
        return mMessageLiveData;
    }
}

File: app/src/main/java/com/vladimir_x/easyenglishlearn/exercises/ExerciseViewModel.kt
package com.vladimir_x.easyenglishlearn.exercises

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import com.vladimir_x.easyenglishlearn.State
import com.vladimir_x.easyenglishlearn.State.CompletedState
import com.vladimir_x.easyenglishlearn.State.DataState
import com.vladimir_x.easyenglishlearn.State.ErrorState
import com.vladimir_x.easyenglishlearn.model.Answer
import com.vladimir_x.easyenglishlearn.model.Word

abstract class ExerciseViewModel : ViewModel() {
    private var iteration = 0
    private var errorCount = 0
    private var question: String = ""
    var translationDirection = true
    var currentWord: Word? = null
    val wordList = mutableListOf<Word>()

    private val _exerciseState = MutableLiveData<State>(State.IdleState)
    val exerciseState: LiveData<State>
        get() = _exerciseState

    private val isExerciseOver: Boolean
        get() = iteration >= wordList.size

    open fun prepareQuestionAndAnswers() {
        currentWord = wordList[iteration].also {
            question = if (translationDirection) it.lexeme else it.translation
        }
    }

    /**
     * This method must be called from fragment when it is first started
     */
    fun startExercise(wordList: List<Word>, translationDirection: Boolean) {
        this.wordList.addAll(wordList)
        this.translationDirection = translationDirection
        errorCount = 0
        iteration = 0
        prepareQuestionAndAnswers()
    }

    fun convertWordToQuestion(word: Word?): String =
        if (translationDirection) word?.lexeme ?: "" else word?.translation ?: ""

    fun checkAnswer(answer: CharSequence) {
        if (isAnswerCorrect(answer)) {
            iteration++
            if (isExerciseOver) {
                finishExercise()
            } else {
                prepareQuestionAndAnswers()
            }
        } else {
            errorCount++
            showError()
            prepareQuestionAndAnswers()
        }
    }

    fun sendData(data: DataDto) {
        changeState(DataState(data))
    }

    private fun changeState(state: State) {
        _exerciseState.value = state
    }

    private fun finishExercise() {
        changeState(CompletedState(errorCount))
    }

    private fun showError() {
        changeState(ErrorState())
    }

    private fun isAnswerCorrect(answer: CharSequence): Boolean =
        Answer(currentWord, answer, translationDirection).isCorrect
}